### DISCUSSION TITLE
GSM SMS message Extracting using Arduino

### DISCUSSION THREAD

**Discussion by josephnyamedi** [community/community#142961] [Programming Help]
### Select Topic Area

General

### Body

Please I'm student and I help on a project that I want to extract some parts SMS message received on GSM module Using Arduino. Please how should I Go about it with the coding part?

**Comment by thomassiard** [comment]
Step 1: Set Up the GSM Module
First, ensure that your GSM module is properly connected to the Arduino. Most GSM modules communicate over Serial, so you’ll use Serial or SoftwareSerial if you need a second serial port.

Step 2: Initialize GSM and Read SMS
Below is an example code to initialize the GSM module, read an SMS, and extract specific parts of the message:

#include <SoftwareSerial.h>
SoftwareSerial gsm(7, 8); // RX, TX pins

void setup() {
  Serial.begin(9600);   // Serial monitor
  gsm.begin(9600);      // GSM module baud rate
  delay(1000);

  // Send AT commands to the GSM module to initialize SMS mode
  gsm.println("AT");                // Check connection
  delay(1000);
  gsm.println("AT+CMGF=1");         // Set SMS to text mode
  delay(1000);
  gsm.println("AT+CNMI=1,2,0,0,0"); // Configure to display SMS directly
}

void loop() {
  if (gsm.available()) {
    String message = gsm.readString(); // Read the SMS message
    Serial.println("Received Message: ");
    Serial.println(message);

    // Step 3: Extract specific part of the SMS
    extractPart(message, "Keyword:", "\n");
  }
}

// Function to extract text between two delimiters
void extractPart(String message, String startDelimiter, String endDelimiter) {
  int startIndex = message.indexOf(startDelimiter);
  int endIndex = message.indexOf(endDelimiter, startIndex + startDelimiter.length());

  if (startIndex != -1 && endIndex != -1) {
    String extractedText = message.substring(startIndex + startDelimiter.length(), endIndex);
    Serial.print("Extracted Text: ");
    Serial.println(extractedText);
  } else {
    Serial.println("Delimiter not found");
  }
}

Explanation of the Code
Initialization:

The gsm.println("AT+CMGF=1"); command sets the GSM module to SMS text mode.
AT+CNMI=1,2,0,0,0 tells the module to display incoming SMS messages directly on the serial port.
Read Message:

gsm.readString() reads the incoming SMS. You may see headers along with the message text, so make sure to filter those out if needed.
Extract Specific Parts:

The extractPart function takes the full message and extracts text between specified delimiters. Adjust the delimiters (Keyword: and \n in this example) based on the structure of the SMS you’re receiving.
Additional Tips
Test each part step-by-step to make sure you’re receiving messages correctly.
Modify the extractPart function if you need to extract multiple pieces of information from the message.

**Comment by josephnyamedi** [comment]
Thank you so much.

**Comment by josephnyamedi** [comment]
Send number, timestamp, reference number or name. So that I can give this
instruction that if the information are true then operate motor.
